src/ReductionMsg.chpl

```diff
@@ -960,8 +960,10 @@ module ReductionMsg
       const dom = makeDistDom(newSize);
       var ret = makeDistArray(dom, intype);
 
-      forall (oldSegStart, newSegStart, newLength) in zip(segments, newSegs, newSegLengths) with (var agg = newSrcAggregator(intype)){
-          agg.copy(ret[newSegStart..#newLength], values[oldSegStart..#newLength]);
+      forall (oldSegStart, newSegStart, newLength) in zip(segments, newSegs, newSegLengths) with (const ref values, ref ret){
+        forall (r,v) in zip(ret[newSegStart..#newLength], values[oldSegStart..#newLength]) with (var agg = newSrcAggregator(intype)){
+          agg.copy(r, v);
+        }
```

Comment:
My preferred way to do something like this is to use a low level localizing slice to avoid actually slicing the arrays
https://github.com/Bears-R-Us/arkouda/blob/e9023c61ab04e5f160888bd8c8e1e17b60e35207/src/AryUtil.chpl#L541-L549

For copy aggregation, one side has to be local (remote-to-remote aggregation isn't supported). So for this I made a `lowLevelLocalizingSlice` of the source and used a destination aggregator to write into `ret` at a potentially remote index

```chapel
forall (oldSegStart, newSegStart, newLength) in zip(segments, newSegs, newSegLengths) with (ref values,
                                                                                            var agg = newDstAggregator(intype)){
    var v = new lowLevelLocalizingSlice(values, oldSegStart..#newLength);
    for i in 0..#newLength {
      agg.copy(ret[newSegStart+i], v.ptr[i]);
    }
}
```
using a `for` as the inner loop is fine since `newLength` should be small for `head`/`tail`

Reply:
I went ahead and made sure this compiled for multi locale and did some comm testing just to make sure there was a reduction

running the following python code:
```python
   >>>          size = 10**7
   ...:         bool_col = ak.full(size, False, dtype=ak.bool)
   ...:         bool_col[::2] = True
   ...:         df = ak.DataFrame(
   ...:             {
   ...:                 "a": ak.arange(size) % 3,
   ...:                 "b": ak.arange(size, dtype="int64"),
   ...:                 "c": ak.arange(size, dtype="float64"),
   ...:                 "d": ak.random_strings_uniform(size=size, minlen=1, maxlen=2, seed=18),
   ...:                 "e": bool_col,
   ...:             }
   ...:         )
   ...:         df.groupby("a").head(n=2, sort_index=False)
```

comm with this PR:
```
| locale | get | put | execute_on | execute_on_nb |
| -----: | --: | --: | ---------: | ------------: |
|      0 |  43 |  12 |          0 |           132 |
|      1 | 103 |  13 |          6 |            39 |
|      2 | 103 |  14 |         10 |            39 |
|      3 | 119 |  14 |          6 |             0 |
```

comm using localizing slices:
```
| locale | get | put | execute_on | execute_on_nb |
| -----: | --: | --: | ---------: | ------------: |
|      0 |  11 |   0 |          0 |            93 |
|      1 |  77 |   8 |          4 |             0 |
|      2 |  69 |   9 |          4 |             0 |
|      3 |  62 |   8 |          2 |             0 |
```

Reply:
I made the change since you're version seemed better.